lib/Sema/CSApply.cpp

```diff
@@ -7066,11 +7087,54 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
                                             /*isImplicit*/ true));
   }
 
-  case TypeKind::Pack:
-  case TypeKind::PackExpansion: {
+  case TypeKind::Pack: {
     llvm_unreachable("Unimplemented!");
   }
 
+  case TypeKind::PackExpansion: {
+    auto toExpansionType = toType->getAs<PackExpansionType>();
+    auto *expansion = dyn_cast<PackExpansionExpr>(expr);
+
+    // FIXME: Duplicated code from `PackReferenceFinder` in PreCheckExpr.
+    auto toElementType = toExpansionType->getPatternType();
+    toElementType =
+        toElementType->mapTypeOutOfContext().transform([&](Type type) -> Type {
+          auto *genericParam = type->getAs<GenericTypeParamType>();
+          if (!genericParam || !genericParam->isParameterPack())
+            return type;
+
+          auto param = GenericTypeParamType::get(/*isParameterPack*/false,
+                                                 genericParam->getDepth(),
+                                                 genericParam->getIndex(), ctx);
+          return expansion->getGenericEnvironment()->mapTypeIntoContext(param);
+        });
+
+    auto *pattern = coerceToType(expansion->getPatternExpr(),
+                                 toElementType, locator);
+
+    // FIXME: Duplicated code from `visitPackExpansionExpr` in CSGen.
+    auto patternType =
+        cs.getType(pattern).transform([&](Type type) -> Type {
+          auto *element = type->getAs<ElementArchetypeType>();
+          if (!element)
+            return type;
+
+          auto *elementParam = element->mapTypeOutOfContext()->getAs<GenericTypeParamType>();
+          auto *pack = GenericTypeParamType::get(/*isParameterPack*/true,
+                                                 elementParam->getDepth(),
+                                                 elementParam->getIndex(),
+                                                 ctx);
+          return cs.DC->mapTypeIntoContext(pack);
+        });
+    auto shapeType = toExpansionType->getCountType();
+    auto expansionTy = PackExpansionType::get(patternType, shapeType);
+
+    return cs.cacheType(PackExpansionExpr::create(ctx, pattern,
```

Comment:
Instead of creating new expansion expression it could be updated with new type and pattern in-place?

Reply:
It could, sure, but I'm working on moving `...` resolution into the solver, so `PackExpansionExpr`s will only be created in CSApply anyway

Reply:
Sounds good then!